rdmd.d

```diff
@@ -284,13 +287,78 @@ int main(string[] args)
         lastBuildTime = buildWitness.timeLastModified(SysTime.min);
     }
 
-    // Have at it
-    if (chain(root.only, myDeps.byKey).array.anyNewerThan(lastBuildTime))
+    // Fetch dependencies
+    const HashSet!DepNode[DepNode] myDeps = getDependencies(root, workDir,
+        compilerFlags, addStubMain);
+
+    // --makedepend mode. Just print dependencies and exit.
+    if (makeDepend)
+    {
+        writeDeps(exe, root, myDeps, stdout);
+        return 0;
+    }
+
+    // --makedepfile mode. Print dependencies to a file and continue.
+    // This is similar to GCC's -MF option, very useful to update the
+    // dependencies file and compile in one go:
+    // -include .deps.mak
+    // prog:
+    //      rdmd --makedepfile=.deps.mak --build-only prog.d
+    if (makeDepFile !is null)
+        writeDeps(exe, root, myDeps, File(makeDepFile, "w"));
+
+    string d2obj(string dfile)
+    {
+        return buildPath(objDir, dfile.relativePath).setExtension(objExt);
+    }
+
+    /*
+    Put together file arguments for the compile and link commands.
+    For source files, decide if to (re-)compile or if to reuse an existing
+    object file.
+    Also pick out the libraries. They need to be passed explicitly in the
+    link command.
+    */
+    const SysTime[string] times = !force ? getTimes(myDeps) :
+        (SysTime[string]).init;
+    HashSet!string toCompile;
+    HashSet!string toReuse;
+    HashSet!string libraries;
+    foreach (node; myDeps.byKey)
+    {
+        final switch (node.type)
+        {
+            case DepNode.Type.source:
+            immutable o = d2obj(node.file);
+            if (force || times[node.file].newerThan(o))
```

Comment:
Line 322 initializes `times` to an empty associative array if `force` is true. Accesses would crash.